Update README.md (#202)
DataDog/dogstatsd-ruby: README.md

### Migrating from v4.x to v5.x

If you are already using DogStatsD-ruby v4.x and you want to migrate to a version v5.x, the major
change concerning you is the new [threading model](#threading-model):

In practice, it means two things:

1. Now that the client is buffering metrics before sending them, you have to call `Datadog::Statsd#flush(sync: true)` if you want synchronous behavior. In most cases, this is not needed, as the companion thread will automatically flush the buffered metrics if the buffer gets full or when you are closing the instance.

2. You have to make sure you are either:

>   * Using a singleton instance of the DogStatsD client instead of creating a new instance whenever you need one; this will let the buffering mechanism flush metrics regularly

Reviewer: I think the note about `fork` here is still appropriate

Author: I thought https://github.com/DataDog/dogstatsd-ruby/pull/199 was already in, but indeed it's not yet. Since I'll be out for a few days anyway, I propose that we wait to merge this PR until that other one is merged.

Reviewer: Makes sense - our main dev for this library (@remeh) is out until Monday anyways so it would be good for him to take a look as well.

Reviewer 2: Note for myself: mergeable once 5.3.0 live.

2. You have to make sure you are either:

  * using singletons instances of the DogStatsD client and not allocating one each time you need one, letting the buffering mechanism flush metrics, it's still a bad solution if the process later forks (see related section below). Or,
  * properly closing your DogStatsD client instance when it is not needed anymore using the method `Datadog::Statsd#close` to release the resources used by the instance and to close the socket
  * Using a singleton instance of the DogStatsD client instead of creating a new instance whenever you need one; this will let the buffering mechanism flush metrics regularly
  * Or properly disposing of the DogStatsD client instance when it is not needed anymore using the method `Datadog::Statsd#close`

If you have issues with the companion thread or the buffering mode, you can instantiate a client that behaves exactly as in v4.x (i.e. no companion thread and flush on every metric submission):

```ruby
# Import the library

> require 'datadog/statsd'

Reviewer: This is still good for this simple example as you can copy/paste it into irb

Author: I don't mind re-adding it but... don't you think it distracts a bit from the point of the example? 🤔  

This being an example for `single_thread: true, buffer_max_pool_size: 1`, it seems really verbose -- I'm guessing if you got this far, you probably already know the basics, since this is advanced user stuff.

Reviewer: If this was an example that was partial in lifecycle (e.g. only showing metrics sending) and/or not showing initialization/instantiation of the class, I would agree with you on the superfluous argument but it does seem like we're depicting the full scope (other than the metrics sending) in here, making it slightly more leaning for me towards having that line in there. I don't feel too strongly about it so it's fine either way.

```ruby
# Create a DogStatsD client instance using UDS
statsd = Datadog::Statsd.new(socket_path: '/path/to/socket/file', single_thread: true, buffer_max_pool_size: 1)
# ...
statsd.close()
```

### v5.x Common Pitfalls

Version v5.x of `dogstatsd-ruby` is using a companion thread for flushing. This provides better performance, but you need to consider the following pitfalls:

> 1. Applications that use `fork` after having created the dogstatsd instance: the child process will need a new dogstatsd instance, as otherwise it will be unable to send metrics.

Reviewer: The reasoning why this is the case is still very useful I think

Author: As a user of the library, how relevant is it for me to learn about this in the "common pitfalls" section? This seems to go really deep into library development details, which I'm not entirely sure belong in this section of the README.

Reviewer: It's a really big and likely pitfall. Most Ruby deployments use `puma` or `unicorn` that `fork()` the processes so having that note in a different part of the readme I think would cause a lot of heartburn for the users that won't be obvious until long after their code is deployed.

Reviewer 2: Not true anymore since 5.3.0, I have to remove that line.

> 2. Applications that create multiple instances of the client without closing them: it is important to `#close` the instance to free the thread and the socket it is using or it will lead to resources leaking (threads and sockets).

Reviewer: ```suggestion
2. Applications that create multiple instances of the client without closing them: it is important to `#close` all instances to free the thread and the socket they are using otherwise you will leak those resources.
```

If you are using [Sidekiq](https://github.com/mperham/sidekiq), please make sure to close the client instances that are instantiated. [See this example on using DogStatsD-ruby v5.x with Sidekiq](https://github.com/DataDog/dogstatsd-ruby/blob/master/examples/sidekiq_example.rb).

If you are using [Puma](https://github.com/puma/puma) or [Unicorn](https://yhbt.net/unicorn.git), please make sure to create the instance of DogStatsD in the workers, not in the main process before it forks to create its workers. See [this comment for more details](https://github.com/DataDog/dogstatsd-ruby/issues/179#issuecomment-845570345).

> Applications that run into issues but can't apply these recommendations should use the `single_thread` mode which avoids the use of the compainion thread.

Reviewer: ```suggestion
Applications that run into issues but can't apply these recommendations should use the `single_thread` mode which disables the use of the companion thread.
```

Author: Ahaha oops, thanks for the catch!

If you are using [Sidekiq](https://github.com/mperham/sidekiq), please make sure to close the client instances that are instantiated. [See this example on using DogStatsD-ruby v5.x with Sidekiq](https://github.com/DataDog/dogstatsd-ruby/blob/master/examples/sidekiq_example.rb).

If you are using [Puma](https://github.com/puma/puma) or [Unicorn](https://yhbt.net/unicorn.git), please make sure to create the instance of DogStatsD in the workers, not in the main process before it forks to create its workers. See [this comment for more details](https://github.com/DataDog/dogstatsd-ruby/issues/179#issuecomment-845570345).

Applications that are in these situations but can't apply these recommendations should enable the `single_thread` mode which does not use a companion thread. Here is how to instantiate a client in this mode:
Applications that run into issues but can't apply these recommendations should use the `single_thread` mode which avoids the use of the compainion thread.
Here is how to instantiate a client in this mode:

```ruby
# Import the library

> require 'datadog/statsd'

Reviewer: Useful for irb

Author: See my question above about this

Reviewer 2: I've no strong opinion on this, I'll merge your recommandation.

After the client is created, you can start sending Service Checks to Datadog. See the dedicated [Service Check Submission: DogStatsD documentation](https://docs.datadoghq.com/developers/service_checks/dogstatsd_service_checks_submission/?tab=ruby) to see how to submit a Service Check to Datadog.

### Maximum packet size in high-throughput scenarios

In order to have the most efficient use of this library in high-throughput scenarios,
recommended values for the maximum packet size have already been set for both UDS (8192 bytes)
and UDP (1432 bytes).

However, if are in control of your network and want to use a different value for the maximum packet
size, you can do it by setting the `buffer_max_payload_size` parameter:

```ruby
statsd = Datadog::Statsd.new('localhost', 8125, buffer_max_payload_size: 4096)
# ...
statsd.close()
```

## Threading model

> Starting with version 5.0, dogstatsd-ruby employs a new threading model where one instance of `Datadog::Statsd` can be shared between threads and where data sending is non-blocking (asynchronous).

Reviewer: ```suggestion
Starting with version 5.0, `dogstatsd-ruby` employs a new threading model where one instance of `Datadog::Statsd` can be shared between threads and where data sending is non-blocking (asynchronous).
```

When you instantiate a `Datadog::Statsd`, a companion thread is spawned. This thread will be called the Sender thread, as it is modeled by the [Sender](../lib/datadog/statsd/sender.rb) class. You can make use of `single_thread: true` to disable this behavior.

> This thread is stopped when you close the statsd client (`Datadog::Statsd#close`). Allocating a lot of statsd clients without calling `#close` after they are not needed anymore can lead to threads being leaked.

Reviewer: ```suggestion
This thread is stopped when you close the statsd client (`Datadog::Statsd#close`). Instantiation a lot of statsd clients without calling `#close` after they are not needed anymore will most likely lead to threads being leaked.
```

Author: Should this be "Instantiating a" or Instantiation of"?

Reviewer: The former - good catch!

Reviewer 2: I'll add the of in the commit removing L78

end while
```

There are three different kinds of messages:

1. a control message to flush the buffer in the connection
2. a control message to synchronize any thread with the sender thread
3. a message to append to the buffer

> There is also an implicit message which is closing the queue which will cause the sender thread to finish processing and exit.

Reviewer: ```suggestion
There is also an implicit message which closes the queue which will cause the sender thread to finish processing and exit.
```

### Usual workflow

You push metrics to the statsd client which writes them quickly to the sender message queue. The sender thread receives those message, buffers them and flushes them to the connection when the buffer limit is reached.

### Flushing

> When calling flush, a specific control message (`:flush`) is sent to the sender thread. When finding it, it flushes its internal buffer into the connection.

Reviewer: ```suggestion
When calling `Datadog::Statsd#flush`, a specific control message (`:flush`) is sent to the sender thread. When the sender thread receives it, it flushes its internal buffer into the connection.
```

### Rendez-vous

> It is possible to ensure a message has been consumed by the sender thread and written to the buffer by simply calling a rendez-vous right after. This is done when you are doing a synchronous flush: using `Datadog::Statsd#flush(sync: true)`.

Reviewer: ```suggestion
It is possible to ensure a message has been consumed by the sender thread and written to the buffer by simply calling a rendez-vous right after. This is done when you are doing a synchronous flush using `Datadog::Statsd#flush(sync: true)`.
```

> Doing so means the caller thread is blocked waiting until the data has been flushed.

Reviewer: ```suggestion
Doing so means the caller thread is blocked and waiting until the data has been flushed by the sender thread.
```
